tools/qlog/src/lib.rs

```diff
@@ -920,20 +923,258 @@ pub enum EventType {
+            EventType::RecoveryEventType(RecoveryEventType::MetricsUpdated) =>
+                "recovery:metrics_updated",
+            EventType::RecoveryEventType(
+                RecoveryEventType::CongestionStateUpdated,
+            ) => "recovery:congestion_state_updated",
+            EventType::RecoveryEventType(RecoveryEventType::LossTimerUpdated) =>
+                "recovery:loss_timer_updated",
+            EventType::RecoveryEventType(RecoveryEventType::PacketLost) =>
+                "recovery:packet_lost",
+            EventType::RecoveryEventType(
+                RecoveryEventType::MarkedForRetransmit,
+            ) => "recovery:marked_for_retransmit",
+
+            EventType::Http3EventType(Http3EventType::ParametersSet) =>
+                "http:parameters_set",
+            EventType::Http3EventType(Http3EventType::StreamTypeSet) =>
+                "http:stream_type_set",
+            EventType::Http3EventType(Http3EventType::FrameCreated) =>
+                "http:frame_created",
+            EventType::Http3EventType(Http3EventType::FrameParsed) =>
+                "http:frame_parsed",
+            EventType::Http3EventType(Http3EventType::PushResolved) =>
+                "http:push_resolved",
+
+            EventType::QpackEventType(QpackEventType::StateUpdated) =>
+                "qpack:state_updated",
+            EventType::QpackEventType(QpackEventType::StreamStateUpdated) =>
+                "qpack:stream_state_updated",
+            EventType::QpackEventType(QpackEventType::DynamicTableUpdated) =>
+                "qpack:dynamic_table_updated",
+            EventType::QpackEventType(QpackEventType::HeadersEncoded) =>
+                "qpack:headers_encoded",
+            EventType::QpackEventType(QpackEventType::HeadersDecoded) =>
+                "qpack:headers_decoded",
+            EventType::QpackEventType(QpackEventType::InstructionCreated) =>
+                "qpack:instruction_created",
+            EventType::QpackEventType(QpackEventType::InstructionParsed) =>
+                "qpack:instruction_parsed",
+
+            _ => unimplemented!(),
+        };
+
+        write!(f, "{}", v)
+    }
+}
+
 #[derive(Serialize, Clone)]
-#[serde(untagged)]
-#[allow(clippy::large_enum_variant)]
-pub enum EventField {
-    RelativeTime(f32),
+pub enum TimeFormat {
+    Absolute,
+    Delta,
+    Relative,
+}
+
+#[serde_with::skip_serializing_none]
+#[derive(Serialize, Clone)]
+pub struct Event {
+    pub time: f32,
+    pub name: String,
+    pub data: EventData,
 
-    Category(EventCategory),
+    pub protocol_type: Option<String>,
+    pub group_id: Option<String>,
 
-    Event(EventType),
+    pub time_format: Option<TimeFormat>,
 
-    Data(EventData),
+    #[serde(skip)]
+    ty: EventType,
 }
 
-#[derive(Serialize, Clone)]
+impl Event {
+    /// Returns a new `Event` object with the provided data at the default time.
+    pub fn new(data: EventData) -> Self {
+        Self::with_time(0.0, data)
```

Comment:
It looks like in most cases the `time` field is set to `0`, so what is its purpose?

Reply:
Agree it looks a bit weird. The `time` field is super important for qlog. I started with it as mandatory param but realised the quiche code handling of events meant that the time value was always 0 to start and then always updated later. On reflection, I think we can probably tweak quiche's behaviour to just construct the Event with the time on a single pass. I'll go back and tweak things in this PR.

Reply:
Please take a look at the new commit I just pushed. I've refactored things a bit to make the code for building Events less convoluted as a user.

Reply:
So, it seems like the `add_event_with_instant()` calls were replaced with `add_event()`, but then all the relative time calculation code is duplicated in quiche itself, why?

Reply:
The main benefit is that it avoids having to pass a mutable object to be updated inside the qlog crate. 

A secondary benefit is that it gives flexibility to create events at arbitrary timepoints. Which could help us log events (server listening, etc) that occur before before the QUIC connection proper.